(feat) Add BodyReader and JsonBodyParser. Fixes iron/iron#287

s-panferov opened the pull request:
Hello @reem. I have had some time to think about iron/iron#287 and I found out that we can't use the current plugin infrastructure because it doesn't allow to report errors. I think that it is bad for end-user. So I decided to rewrite the whole `body_parser` to be a library that just helps to read body into String with proper `limit`.

If we really want this to be a plugin we need to make plugins return `Result`. Note that I had to copy LimitReader because the default one doesn't report errors. 

I'd like to write some tests for this PR, but at first I need to hear your opinion.

s-panferov commented:
ping @reem

reem commented:
I'm actually working on a consistent way to implement configuration for Plugins.

Do you think the plugin return type should be Result?

s-panferov commented:
> Do you think the plugin return type should be Result?

Yes, I do. Otherwise user has no way to know about errors. Even in this small library there are:

* UTF-8 decoding error;
* stream errors;
* body limit error;

I think it's important to know about them and report them back to the client.

reem commented:
Agreed. I will work on the refactor.

reem commented:
I've updated plugins to use Result, try it out.

s-panferov commented:
Ok, I'll try to refactor this today.

s-panferov commented:
@reem please take a look at https://github.com/reem/rust-plugin/issues/8

s-panferov commented:
I have just made refactoring to plugins. One thing that I don't like is that we can't configure the max body length. I created the `BodyMaxLength` key for that, but the only way to set it is to use a middleware :expressionless:.

reem commented:
@s-panferov I imagined for a while that `persistent::Read` would be used to configure plugins. Your thoughts?

reem commented on Cargo.toml:
```diff
@@ -11,4 +11,9 @@ authors = ["Patrick Tran <patrick.tran06@gmail.com>",
 
 [dependencies]
 iron = "*"
+typemap = "*"
```
exported as `iron::TypeMap`

reem commented on Cargo.toml:
```diff
@@ -11,4 +11,9 @@ authors = ["Patrick Tran <patrick.tran06@gmail.com>",
 iron = "*"
+typemap = "*"
+rustc-serialize = "*"
+error = "*"
```
relevant bits should be exported in Iron but I could be wrong here

reem commented on README.md:
```diff
@@ -1,41 +1,31 @@
+        None => println!("Invalid or no body!"),
     }
-    Ok(Response::with(status::Ok, ""))
+    Ok(iron::Response::with(status::Ok))
```
use the prelude and remove all these `iron::`

reem commented on src/lib.rs:
```diff
@@ -1,37 +1,81 @@
+                })
+            }
+        },
+        Err(err) => Err(errors::BodyError{
```
space before { here

reem commented on src/lib.rs:
```diff
@@ -1,37 +1,81 @@
+        Ok(bytes) => {
+             match String::from_utf8(bytes) {
+                Ok(e) => Ok(e),
+                Err(err) => Err(errors::BodyError{
```
space before { at EOL

reem commented on src/lib.rs:
```diff
@@ -1,37 +1,81 @@
+        let need_read = {
+            let header = req.headers.get::<iron::headers::ContentType>();
+            match header {
+                Some(header) => match **header {
```
@reem wishes we had `matches!` (`headers.get(..).map(|header| matches!(**header, mime::Mime(..)))` instead of this whole mess)

reem commented on src/lib.rs:
```diff
@@ -1,37 +1,81 @@
+        };
+
+        if need_read {
+            let max_length = req.extensions.get::<BodyMaxLength>().map(|val| val.clone()).unwrap_or(DEFAULT_BODY_LIMIT);
```
line too long, use `.cloned()` and indent/split it into two lines

reem commented on src/limit_reader.rs:
```diff
@@ -0,0 +1,40 @@
+use std::cmp;
+
+/// This impl was copied from https://github.com/rust-lang/rust/blob/17bc7d8d5be3be9674d702ccad2fa88c487d23b0/src/libstd/old_io/util.rs#L20
+/// We need this copy to return an error properly.
```
Can you just elaborate on this a bit more?

reem commented:
I think parts of plugin are exported from Iron too, check if it has what you need.

s-panferov commented:
@reem I have addressed all the comments. Also I tried this with curl, everything seems working. I think that Iron must ship something like [this](https://github.com/rustless/rustless/blob/master/src/backend/simple_request.rs) to build test requests.

> @s-panferov I imagined for a while that persistent::Read would be used to configure plugins. Your thoughts?

I haven't worked with it yet, so I can't answer. The docs doesn't say to me much. But I looked at the source. There are a couple of locks and I think it is not one wants to use for _read-only_ configuration on high-load instances.

reem commented:
@s-panferov The Read variant is equivalent to `Arc` in efficiency, with no locking of any kind, only atomic operations and fences.

s-panferov commented:
@reem Ok, but to try it we need to fix it first. I have tried, but there is a recursive problem with new Result type in Plugins :disappointed:. You need to look.

hugoduncan commented:
https://github.com/iron/urlencoded/pull/46 will need updating to use this when it lands.

s-panferov commented:
@reem, the second commit with `persistent` was added. Re-r please.

reem commented on README.md:
```diff
@@ -1,41 +1,39 @@
+fn log_body(req: &mut iron::Request) -> iron::IronResult<iron::Response> {
+    let parsed = req.get::<bodyparser::BodyReader>();
+    match parsed {
+        Ok(Some(body)) => println!("Parsed body:\n{}", body),
```
It's not exactly parsed, is it?

reem commented on README.md:
```diff
@@ -1,41 +1,39 @@
+use plugin::Pluggable;
+use persistent::Read;
 
-use iron::{Iron, Request, Response, IronResult, Plugin, status};
```
use `iron::prelude::*` instead of a prepending everything with `iron::`

reem commented on README.md:
```diff
@@ -1,41 +1,39 @@
+extern crate persistent;
 
-use std::io::net::ip::Ipv4Addr;
+use plugin::Pluggable;
```
this is in the iron prelude, no need to import it or depend on plugin as a user

reem commented on examples/bodyparser.rs:
```diff
@@ -1,31 +1,30 @@
-#![feature(globs)]
 extern crate iron;
```
All the same comments to the copy of the example here.

s-panferov commented:
@reem Hello, I have pushed fixes to your comments (second commit was amended). Also I have pushed one more commit with two parsers, that work on top of the `BodyReader`.

I think that I have done all what I want in this lib, so please re-r.

reem commented on examples/bodyparser.rs:
```diff
@@ -1,31 +1,52 @@
-    age: Option<i8>,
+#[derive(Debug)]
+#[derive(RustcDecodable)]
+#[derive(Clone)]
```
Can these all go on one `derive` line?

reem commented on examples/bodyparser.rs:
```diff
@@ -1,31 +1,52 @@
+fn log_body(req: &mut Request) -> IronResult<Response> {
+    let body = req.get::<bodyparser::BodyReader>();
+    match body {
+        Ok(Some(body)) => println!("Readed body:\n{}", body),
```
Did you mean `Read body`?

reem commented on src/lib.rs:
```diff
@@ -1,37 +1,139 @@
 #![crate_name = "bodyparser"]
-#![feature(default_type_params)]
+#![feature(core)]
```
Can go in one feature directive

s-panferov commented:
@reem done

reem commented on examples/bodyparser.rs:
```diff
@@ -1,31 +1,50 @@
+        Err(err) => println!("Error: {:?}", err)
+    }
+
+    let json_body = req.get::<bodyparser::JsonBodyParser>();
```
I think this is a bit too verbose. Could you remove the `BodyParser` suffix from all these types and rename `BodyReader` to `Raw`?

s-panferov replied:
Sure

s-panferov replied:
But maybe we need to think more about `Json`, it will clash with `json::Json`

s-panferov commented:
@reem updated

reem commented:
LGTM. Awesome job @s-panferov.
